clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -854,7 +856,7 @@ synchronized TxnRequestHandler nextRequest(boolean hasIncompleteBatches) {
             return null;
         }
 
-        if (nextRequestHandler.isEndTxn() && (!isTransactionV2Enabled() && !transactionStarted)) {
+        if (nextRequestHandler.isEndTxn() && !transactionStarted) {
```

Comment:
minor nit here: I wonder if we can change the log here for tv2 so that we say we didn't attempt to add any partitions.

Reply:
Pardon me, could you give me some hint? Did you mean we need to modify the log in line 862
```java
                log.debug("Not sending EndTxn for completed transaction since no partitions " +
                        "or offsets were successfully added");
```
to something else when using TV2?

Reply:
Right -- the implication of successfully added is a little confusing in the TV2 case because it is not a matter of "successfully" adding but a matter of not attempting to add any at all.

Reply:
Got it, I modified the log to `Not sending EndTxn for completed transaction since no send or sendOffsetsToTransaction were triggered` under TV2. Thank you.